core/metacling/src/TClingCallbacks.cxx

```diff
@@ -783,3 +785,11 @@ void TClingCallbacks::LibraryUnloaded(const void* dyLibHandle,
 void TClingCallbacks::PrintStackTrace() {
    TCling__PrintStackTrace();
 }
+
+void TClingCallbacks::EnteringUserCode() {
+  TCling__UnlockInterpreterMutex();
+}
+
+void TClingCallbacks::ReturnedFromUserCode() {
+  TCling__LockInterpreterMutex();
```

Comment:
The unlocking ought to be conditional upon the existence of the mutex when we entered and/or (which should be equivalent) on whether we unlocked it.